File: internal/tools/cashflow/cashflow.go
package cashflow

import (
	"encoding/json"
	"fmt"
	"log"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"time"
)

type TransactionType string

const (
	Income  TransactionType = "income"
	Expense TransactionType = "expense"
)

type CurrencyType string

const (
	IDR CurrencyType = "IDR"
	USD CurrencyType = "USD"
	EUR CurrencyType = "EUR"
	JPY CurrencyType = "JPY"
	GBP CurrencyType = "GBP"
)

type Category struct {
	ID   string `json:"id"`
	Name string `json:"name"`
}

type Transaction struct {
	ID          string          `json:"id"`
	UserID      string          `json:"user_id"`
	Type        TransactionType `json:"type"`
	Amount      float64         `json:"amount"`
	Currency    CurrencyType    `json:"currency"`
	Category    Category        `json:"category"`
	Description string          `json:"description"`
	Date        time.Time       `json:"date"`
}

type CashFlowTool struct {
	dataPath string
	mu       sync.RWMutex
}

type CashFlowData struct {
	Transactions []Transaction `json:"transactions"`
	Categories   []Category    `json:"categories"`
}

func NewCashFlowTool() *CashFlowTool {
	workingDir, err := os.Getwd()
	if err != nil {
		log.Printf("Error getting working directory: %v\n", err)
		return nil
	}

	dataDir := filepath.Join(workingDir, "data", "cashflow")
	if err := os.MkdirAll(dataDir, 0755); err != nil {
		log.Printf("Error creating cashflow directory: %v\n", err)
		return nil
	}

	dataPath := filepath.Join(dataDir, "cashflow.json")
	return &CashFlowTool{
		dataPath: dataPath,
	}
}

func (ct *CashFlowTool) loadData() (*CashFlowData, error) {
	ct.mu.RLock()
	defer ct.mu.RUnlock()

	data := &CashFlowData{
		Transactions: make([]Transaction, 0),
		Categories:   make([]Category, 0),
	}

	if _, err := os.Stat(ct.dataPath); err == nil {
		file, err := os.ReadFile(ct.dataPath)
		if err != nil {
			return nil, fmt.Errorf("error reading data file: %v", err)
		}

		if len(file) == 0 {
			return data, nil
		}

		if err := json.Unmarshal(file, data); err != nil {
			log.Printf("Warning: error parsing data file: %v, returning empty data", err)
			return data, nil
		}
	}

	return data, nil
}

func (ct *CashFlowTool) saveData(data *CashFlowData) error {
	ct.mu.Lock()
	defer ct.mu.Unlock()

	if data == nil {
		data = &CashFlowData{
			Transactions: make([]Transaction, 0),
			Categories:   make([]Category, 0),
		}
	}

	if data.Transactions == nil {
		data.Transactions = make([]Transaction, 0)
	}
	if data.Categories == nil {
		data.Categories = make([]Category, 0)
	}

	file, err := json.MarshalIndent(data, "", "  ")
	if err != nil {
		return fmt.Errorf("error marshaling data: %v", err)
	}

	dir := filepath.Dir(ct.dataPath)
	if err := os.MkdirAll(dir, 0755); err != nil {
		return fmt.Errorf("error creating directory: %v", err)
	}

	if err := os.WriteFile(ct.dataPath, file, 0644); err != nil {
		return fmt.Errorf("error writing data file: %v", err)
	}

	return nil
}

func toLowerCase(s string) string {
	return strings.ToLower(s)
}

func parseDate(dateStr string) (time.Time, error) {
	if t, err := time.Parse(time.RFC3339, dateStr); err == nil {
		return t, nil
	}

	if t, err := time.Parse("2006-01-02", dateStr); err == nil {
		return time.Date(t.Year(), t.Month(), t.Day(), 0, 0, 0, 0, time.Local), nil
	}

	return time.Time{}, fmt.Errorf("invalid date format: %s", dateStr)
}

func (ct *CashFlowTool) CallTool(arguments string) string {
	var params map[string]interface{}
	if err := json.Unmarshal([]byte(arguments), &params); err != nil {
		return fmt.Sprintf("Error parsing arguments: %v", err)
	}

	action, ok := params["action"].(string)
	if !ok {
		return "Error: action not found"
	}

	switch action {
	case "add_transaction":
		return ct.handleAddTransaction(params)
	case "get_transactions":
		return ct.handleGetTransactions(params)
	case "update_transaction":
		return ct.handleUpdateTransaction(params)
	case "delete_transaction":
		return ct.handleDeleteTransaction(params)
	case "get_analytics":
		return ct.handleGetAnalytics(params)
	case "add_category":
		return ct.handleAddCategory(params)
	case "get_categories":
		return ct.handleGetCategories()
	default:
		return fmt.Sprintf("Error: invalid action: %s", action)
	}
}

func (ct *CashFlowTool) handleAddTransaction(params map[string]interface{}) string {
	data, err := ct.loadData()
	if err != nil {
		return fmt.Sprintf("Error loading data: %v", err)
	}

	userID, ok := params["user_id"].(string)
	if !ok || userID == "" {
		return "Error: user_id is required"
	}

	transactionData, ok := params["transaction"].(map[string]interface{})
	if !ok {
		return "Error: invalid transaction data"
	}

	transactionType, ok := transactionData["type"].(string)
	if !ok || (transactionType != "income" && transactionType != "expense") {
		return "Error: invalid transaction type"
	}

	amount, ok := transactionData["amount"].(float64)
	if !ok || amount <= 0 {
		return "Error: transaction amount must be greater than 0"
	}

	currency, ok := transactionData["currency"].(string)
	if !ok {
		currency = string(IDR)
	}
	currencyType := CurrencyType(currency)
	if currencyType != IDR && currencyType != USD && currencyType != EUR && currencyType != JPY && currencyType != GBP {
		return "Error: invalid currency"
	}

	dateStr, ok := transactionData["date"].(string)
	if !ok {
		return "Error: invalid date format"
	}
	date, err := parseDate(dateStr)
	if err != nil {
		return fmt.Sprintf("Error: %v", err)
	}

	categoryData, ok := transactionData["category"].(map[string]interface{})
	if !ok {
		return "Error: invalid category data"
	}

	categoryName, ok := categoryData["name"].(string)
	if !ok || categoryName == "" {
		return "Error: category name cannot be empty"
	}

	categoryName = toLowerCase(categoryName)

	transaction := Transaction{
		ID:          fmt.Sprintf("trx_%d", time.Now().UnixNano()),
		UserID:      userID,
		Type:        TransactionType(transactionType),
		Amount:      amount,
		Currency:    currencyType,
		Description: transactionData["description"].(string),
		Date:        date,
		Category: Category{
			ID:   fmt.Sprintf("cat_%d", time.Now().UnixNano()),
			Name: categoryName,
		},
	}

	categoryExists := false
	for _, c := range data.Categories {
		if c.Name == transaction.Category.Name {
			transaction.Category.ID = c.ID
			categoryExists = true
			break
		}
	}
	if !categoryExists {
		data.Categories = append(data.Categories, transaction.Category)
	}

	data.Transactions = append(data.Transactions, transaction)

	if err := ct.saveData(data); err != nil {
		return fmt.Sprintf("Error saving data: %v", err)
	}

	return fmt.Sprintf("Transaction added successfully with ID: %s", transaction.ID)
}

func (ct *CashFlowTool) handleGetTransactions(params map[string]interface{}) string {
	data, err := ct.loadData()
	if err != nil {
		return fmt.Sprintf("Error loading data: %v", err)
	}

	userID, ok := params["user_id"].(string)
	if !ok || userID == "" {
		return "Error: user_id is required"
	}

	dateRange, ok := params["date_range"].(map[string]interface{})
	if !ok {
		return "Error: invalid date range"
	}

	start, err := parseDate(dateRange["start"].(string))
	if err != nil {
		return fmt.Sprintf("Error: invalid start date format: %v", err)
	}

	end, err := parseDate(dateRange["end"].(string))
	if err != nil {
		return fmt.Sprintf("Error: invalid end date format: %v", err)
	}

	var filteredTransactions []Transaction
	for _, t := range data.Transactions {
		if t.UserID == userID && (t.Date.Equal(start) || t.Date.After(start)) && (t.Date.Equal(end) || t.Date.Before(end)) {
			filteredTransactions = append(filteredTransactions, t)
		}
	}

	result, err := json.Marshal(filteredTransactions)
	if err != nil {
		return fmt.Sprintf("Error: failed to convert data: %v", err)
	}

	if err := ct.saveData(data); err != nil {
		return fmt.Sprintf("Error saving data: %v", err)
	}

	return string(result)
}

func (ct *CashFlowTool) handleUpdateTransaction(params map[string]interface{}) string {
	data, err := ct.loadData()
	if err != nil {
		return fmt.Sprintf("Error loading data: %v", err)
	}

	userID, ok := params["user_id"].(string)
	if !ok || userID == "" {
		return "Error: user_id is required"
	}

	transactionID, ok := params["transaction_id"].(string)
	if !ok {
		return "Error: transaction ID not found"
	}

	transactionData, ok := params["transaction"].(map[string]interface{})
	if !ok {
		return "Error: invalid transaction data"
	}

	transactionType, ok := transactionData["type"].(string)
	if !ok || (transactionType != "income" && transactionType != "expense") {
		return "Error: invalid transaction type"
	}

	amount, ok := transactionData["amount"].(float64)
	if !ok || amount <= 0 {
		return "Error: transaction amount must be greater than 0"
	}

	currency, ok := transactionData["currency"].(string)
	if !ok {
		currency = string(IDR)
	}
	currencyType := CurrencyType(currency)
	if currencyType != IDR && currencyType != USD && currencyType != EUR && currencyType != JPY && currencyType != GBP {
		return "Error: invalid currency"
	}

	dateStr, ok := transactionData["date"].(string)
	if !ok {
		return "Error: invalid date format"
	}
	date, err := parseDate(dateStr)
	if err != nil {
		return fmt.Sprintf("Error: %v", err)
	}

	categoryData, ok := transactionData["category"].(map[string]interface{})
	if !ok {
		return "Error: invalid category data"
	}

	categoryName, ok := categoryData["name"].(string)
	if !ok || categoryName == "" {
		return "Error: category name cannot be empty"
	}

	categoryName = toLowerCase(categoryName)

	found := false
	for i, t := range data.Transactions {
		if t.ID == transactionID && t.UserID == userID {
			categoryExists := false
			for _, c := range data.Categories {
				if c.Name == categoryName {
					data.Transactions[i] = Transaction{
						ID:          transactionID,
						UserID:      userID,
						Type:        TransactionType(transactionType),
						Amount:      amount,
						Currency:    currencyType,
						Description: transactionData["description"].(string),
						Date:        date,
						Category:    c,
					}
					categoryExists = true
					break
				}
			}

			if !categoryExists {
				newCategory := Category{
					ID:   fmt.Sprintf("cat_%d", time.Now().UnixNano()),
					Name: categoryName,
				}
				data.Categories = append(data.Categories, newCategory)
				data.Transactions[i] = Transaction{
					ID:          transactionID,
					UserID:      userID,
					Type:        TransactionType(transactionType),
					Amount:      amount,
					Currency:    currencyType,
					Description: transactionData["description"].(string),
					Date:        date,
					Category:    newCategory,
				}
			}

			found = true
			break
		}
	}

	if !found {
		return "Error: transaction not found"
	}

	if err := ct.saveData(data); err != nil {
		return fmt.Sprintf("Error saving data: %v", err)
	}

	return fmt.Sprintf("Transaction updated successfully with ID: %s", transactionID)
}

func (ct *CashFlowTool) handleDeleteTransaction(params map[string]interface{}) string {
	data, err := ct.loadData()
	if err != nil {
		return fmt.Sprintf("Error loading data: %v", err)
	}

	userID, ok := params["user_id"].(string)
	if !ok || userID == "" {
		return "Error: user_id is required"
	}

	transactionID, ok := params["transaction_id"].(string)
	if !ok {
		return "Error: transaction ID not found"
	}

	found := false
	for i, t := range data.Transactions {
		if t.ID == transactionID && t.UserID == userID {
			data.Transactions = append(data.Transactions[:i], data.Transactions[i+1:]...)
			found = true
			break
		}
	}

	if !found {
		return "Error: transaction not found"
	}

	if err := ct.saveData(data); err != nil {
		return fmt.Sprintf("Error saving data: %v", err)
	}

	return "Transaction deleted successfully"
}

func (ct *CashFlowTool) handleGetAnalytics(params map[string]interface{}) string {
	data, err := ct.loadData()
	if err != nil {
		return fmt.Sprintf("Error loading data: %v", err)
	}

	userID, ok := params["user_id"].(string)
	if !ok || userID == "" {
		return "Error: user_id is required"
	}

	dateRange, ok := params["date_range"].(map[string]interface{})
	if !ok {
		return "Error: invalid date range"
	}

	start, err := parseDate(dateRange["start"].(string))
	if err != nil {
		return fmt.Sprintf("Error: invalid start date format: %v", err)
	}

	end, err := parseDate(dateRange["end"].(string))
	if err != nil {
		return fmt.Sprintf("Error: invalid end date format: %v", err)
	}

	var totalIncome, totalExpense float64
	incomeByCategory := make(map[string]float64)
	expenseByCategory := make(map[string]float64)
	transactionCount := 0

	for _, t := range data.Transactions {
		if t.UserID == userID && (t.Date.Equal(start) || t.Date.After(start)) && (t.Date.Equal(end) || t.Date.Before(end)) {
			transactionCount++
			if t.Type == Income {
				totalIncome += t.Amount
				incomeByCategory[t.Category.Name] += t.Amount
			} else {
				totalExpense += t.Amount
				expenseByCategory[t.Category.Name] += t.Amount
			}
		}
	}

	analytics := map[string]interface{}{
		"total_income":        totalIncome,
		"total_expense":       totalExpense,
		"balance":             totalIncome - totalExpense,
		"income_by_category":  incomeByCategory,
		"expense_by_category": expenseByCategory,
		"transaction_count":   transactionCount,
	}

	result, err := json.Marshal(analytics)
	if err != nil {
		return fmt.Sprintf("Error: failed to convert data: %v", err)
	}

	return string(result)
}

func (ct *CashFlowTool) handleAddCategory(params map[string]interface{}) string {
	data, err := ct.loadData()
	if err != nil {
		return fmt.Sprintf("Error loading data: %v", err)
	}

	categoryData, ok := params["category"].(map[string]interface{})
	if !ok {
		return "Error: invalid category data"
	}

	name, ok := categoryData["name"].(string)
	if !ok || name == "" {
		return "Error: category name cannot be empty"
	}

	name = toLowerCase(name)

	for _, c := range data.Categories {
		if c.Name == name {
			return "Error: category with this name already exists"
		}
	}

	id, _ := categoryData["id"].(string)
	if id == "" {
		id = fmt.Sprintf("cat_%d", time.Now().UnixNano())
	}

	category := Category{
		ID:   id,
		Name: name,
	}

	data.Categories = append(data.Categories, category)

	if err := ct.saveData(data); err != nil {
		return fmt.Sprintf("Error saving data: %v", err)
	}

	return "Category added successfully"
}

func (ct *CashFlowTool) handleGetCategories() string {
	data, err := ct.loadData()
	if err != nil {
		return fmt.Sprintf("Error loading data: %v", err)
	}

	result, err := json.Marshal(data.Categories)
	if err != nil {
		return fmt.Sprintf("Error: failed to convert data: %v", err)
	}

	return string(result)
}

File: .teo/skills/cashflow/scripts/cashflow.py
import sys
import json
import os
import datetime
import time

# Constants
DATA_DIR = os.path.join("data", "cashflow")
DATA_FILE = os.path.join(DATA_DIR, "cashflow.json")

def ensure_data_dir():
    if not os.path.exists(DATA_DIR):
        os.makedirs(DATA_DIR)

def load_data():
    ensure_data_dir()
    if os.path.exists(DATA_FILE):
        try:
            with open(DATA_FILE, 'r') as f:
                data = json.load(f)
                # Ensure structure
                if "transactions" not in data: data["transactions"] = []
                if "categories" not in data: data["categories"] = []
                return data
        except Exception:
            pass
    return {"transactions": [], "categories": []}

def save_data(data):
    ensure_data_dir()
    with open(DATA_FILE, 'w') as f:
        json.dump(data, f, indent=2)

def handle_add_transaction(user_id, params):
    data = load_data()
    t_data = params.get("transaction", {})
    
    # Validation
    if t_data.get("amount", 0) <= 0:
        return json.dumps({"error": "Error: transaction amount must be greater than 0"})
    if t_data.get("type") not in ["income", "expense"]:
        return json.dumps({"error": "Error: invalid transaction type"})
    
    cat_name = t_data.get("category", {}).get("name", "").lower()
    if not cat_name:
         return json.dumps({"error": "Error: category name cannot be empty"})

    # Check/Create Category
    cat_id = f"cat_{time.time_ns()}"
    cat_obj = {"id": cat_id, "name": cat_name}
    
    found_cat = False
    for c in data["categories"]:
        if c["name"] == cat_name:
            cat_obj = c
            found_cat = True
            break
    
    if not found_cat:
        data["categories"].append(cat_obj)

    # Create Transaction
    new_t = {
        "id": f"trx_{time.time_ns()}",
        "user_id": user_id,
        "type": t_data["type"],
        "amount": t_data["amount"],
        "currency": t_data.get("currency", "IDR"),
        "category": cat_obj,
        "description": t_data.get("description", ""),
        "date": t_data.get("date", datetime.datetime.now().isoformat())
    }
    
    data["transactions"].append(new_t)
    save_data(data)
    return json.dumps({"message": f"Transaction added successfully with ID: {new_t['id']}"})

def handle_get_transactions(user_id, params):
    data = load_data()
    date_range = params.get("date_range", {})
    start = date_range.get("start")
    end = date_range.get("end")
    
    if not start or not end:
        return json.dumps({"error": "Error: invalid date range"})

    results = []
    # Simple string comparison for ISO dates
    for t in data["transactions"]:
        if t["user_id"] == user_id:
            if t["date"] >= start and t["date"] <= end:
                results.append(t)
                
    return json.dumps(results, indent=2)

def handle_update_transaction(user_id, params):
    data = load_data()
    trx_id = params.get("transaction_id")
    t_data = params.get("transaction", {})
    
    if not trx_id: return json.dumps({"error": "Error: transaction ID not found"})

    # Find transaction
    trx_idx = -1
    for i, t in enumerate(data["transactions"]):
        if t["id"] == trx_id and t["user_id"] == user_id:
            trx_idx = i
            break
            
    if trx_idx == -1:
        return json.dumps({"error": "Error: transaction not found"})

    # Update logic (similar to add, handle category check)
    cat_name = t_data.get("category", {}).get("name", "").lower()
    if cat_name:
        cat_obj = {"id": f"cat_{time.time_ns()}", "name": cat_name}
        found_cat = False
        for c in data["categories"]:
            if c["name"] == cat_name:
                cat_obj = c
                found_cat = True
                break
        if not found_cat:
            data["categories"].append(cat_obj)
        data["transactions"][trx_idx]["category"] = cat_obj

    if "amount" in t_data: data["transactions"][trx_idx]["amount"] = t_data["amount"]
    if "type" in t_data: data["transactions"][trx_idx]["type"] = t_data["type"]
    if "description" in t_data: data["transactions"][trx_idx]["description"] = t_data["description"]
    if "date" in t_data: data["transactions"][trx_idx]["date"] = t_data["date"]
    
    save_data(data)
    return json.dumps({"message": f"Transaction updated successfully with ID: {trx_id}"})

def handle_delete_transaction(user_id, params):
    data = load_data()
    trx_id = params.get("transaction_id")
    
    initial_len = len(data["transactions"])
    data["transactions"] = [t for t in data["transactions"] if not (t["id"] == trx_id and t["user_id"] == user_id)]
    
    if len(data["transactions"]) == initial_len:
        return json.dumps({"error": "Error: transaction not found"})
        
    save_data(data)
    return json.dumps({"message": "Transaction deleted successfully"})

def handle_get_analytics(user_id, params):
    data = load_data()
    date_range = params.get("date_range", {})
    start = date_range.get("start")
    end = date_range.get("end")
    
    total_income = 0
    total_expense = 0
    income_by_category = {}
    expense_by_category = {}
    count = 0
    
    for t in data["transactions"]:
        if t["user_id"] == user_id:
             if t["date"] >= start and t["date"] <= end:
                 count += 1
                 amt = t["amount"]
                 cat = t["category"]["name"]
                 
                 if t["type"] == "income":
                     total_income += amt
                     income_by_category[cat] = income_by_category.get(cat, 0) + amt
                 else:
                     total_expense += amt
                     expense_by_category[cat] = expense_by_category.get(cat, 0) + amt
                     
    analytics = {
        "total_income": total_income,
        "total_expense": total_expense,
        "balance": total_income - total_expense,
        "income_by_category": income_by_category,
        "expense_by_category": expense_by_category,
        "transaction_count": count
    }
    return json.dumps(analytics, indent=2)

def handle_add_category(params):
    data = load_data()
    cat_data = params.get("category", {})
    name = cat_data.get("name", "").lower()
    
    if not name: return json.dumps({"error": "Error: category name cannot be empty"})
    
    for c in data["categories"]:
        if c["name"] == name:
            return json.dumps({"error": "Error: category with this name already exists"})
            
    new_cat = {
        "id": cat_data.get("id") or f"cat_{time.time_ns()}",
        "name": name
    }
    data["categories"].append(new_cat)
    save_data(data)
    return json.dumps({"message": "Category added successfully"})

def handle_get_categories():
    data = load_data()
    return json.dumps(data["categories"], indent=2)

def main():
    if len(sys.argv) < 2:
        print(json.dumps({"error": "No arguments provided"}))
        return

    try:
        args = json.loads(sys.argv[1])
    except json.JSONDecodeError:
        print(json.dumps({"error": "Invalid JSON arguments"}))
        return

    action = args.get("action", "")
    user_id = args.get("user_id", "")
    
    # Validation for user_id on user-specific actions
    if action in ["add_transaction", "get_transactions", "update_transaction", "delete_transaction", "get_analytics"] and not user_id:
         print(json.dumps({"error": "Error: user_id is required"}))
         return

    if action == "add_transaction":
        print(handle_add_transaction(user_id, args))
    elif action == "get_transactions":
        print(handle_get_transactions(user_id, args))
    elif action == "update_transaction":
        print(handle_update_transaction(user_id, args))
    elif action == "delete_transaction":
        print(handle_delete_transaction(user_id, args))
    elif action == "get_analytics":
        print(handle_get_analytics(user_id, args))
    elif action == "add_category":
        print(handle_add_category(args))
    elif action == "get_categories":
        print(handle_get_categories())
    else:
        print(json.dumps({"error": f"Error: invalid action: {action}"}))

if __name__ == "__main__":
    main()
